pkg/router/template/router.go

```diff
@@ -386,18 +386,20 @@ func (r *templateRouter) readState() error {
 // the state and refresh the backend. This is all done in the background
 // so that we can rate limit + coalesce multiple changes.
 func (r *templateRouter) Commit() {
+
 	r.lock.Lock()
-	defer r.lock.Unlock()
 
 	if !r.synced {
 		glog.V(4).Infof("Router state synchronized for the first time")
 		r.synced = true
 		r.stateChanged = true
 	}
 
-	if r.stateChanged {
-		r.rateLimitedCommitFunction.Invoke(r.rateLimitedCommitFunction)
-		r.stateChanged = false
+	needsCommit := r.stateChanged
+	r.lock.Unlock()
+
+	if needsCommit {
+		go r.rateLimitedCommitFunction.Invoke(r.rateLimitedCommitFunction)
```

Comment:
Nope... stupid idea.  It blocks earlier because the long-running job holds the router lock.  Pulling the change.